rust/processor/src/models/ans_models/ans_lookup.rs

```diff
@@ -0,0 +1,776 @@
+                },
+            }
+        }
+        None
+    }
+
+    fn lookup_name_record_by_domain_and_subdomain(
+        domain: &str,
+        subdomain: &str,
+        current_ans_lookups: &HashMap<CurrentAnsLookupPK, Self>,
+        conn: &mut PgPoolConnection,
+    ) -> Option<Self> {
+        // Check if the domain is in current_ans_lookups
+        let maybe_name_record =
+            current_ans_lookups.get(&(domain.to_string(), subdomain.to_string()));
+        if let Some(name_record) = maybe_name_record {
+            return Some(name_record.clone());
+        }
+
+        let mut retried = 0;
+        while retried < QUERY_RETRIES {
+            retried += 1;
+            match CurrentAnsLookupDataQuery::get_name_record_by_domain_and_subdomain(
+                conn, domain, subdomain,
+            ) {
+                Ok(ans_record) => {
+                    return Some(CurrentAnsLookup {
+                        domain: ans_record.domain,
+                        subdomain: ans_record.subdomain,
+                        registered_address: ans_record.registered_address,
+                        last_transaction_version: ans_record.last_transaction_version,
+                        expiration_timestamp: ans_record.expiration_timestamp,
+                        token_name: ans_record.token_name,
+                        is_primary: ans_record.is_primary,
+                        is_deleted: ans_record.is_deleted,
+                    });
+                },
+                Err(_) => {
+                    std::thread::sleep(std::time::Duration::from_millis(QUERY_RETRY_DELAY_MS));
+                },
+            }
+        }
+        None
+    }
+}
+
+impl CurrentAnsLookupDataQuery {
+    pub fn get_name_record_by_domain_and_subdomain(
+        conn: &mut PgPoolConnection,
+        domain: &str,
+        subdomain: &str,
+    ) -> diesel::QueryResult<Self> {
+        current_ans_lookup::table
+            .filter(current_ans_lookup::domain.eq(domain))
+            .filter(current_ans_lookup::subdomain.eq(subdomain))
+            .first::<Self>(conn)
+    }
+
+    pub fn get_name_record_by_primary_address(
+        conn: &mut PgPoolConnection,
+        registered_address: &str,
+    ) -> diesel::QueryResult<Self> {
+        current_ans_lookup::table
+            .filter(current_ans_lookup::registered_address.eq(registered_address))
+            .filter(current_ans_lookup::is_primary.eq(true))
+            .first::<Self>(conn)
+    }
+}
+
+impl OptionalString {
+    fn get_string(&self) -> Option<String> {
+        if self.vec.is_empty() {
+            None
+        } else {
+            Some(self.vec[0].clone())
+        }
+    }
+}
+
+impl NameRecordKeyV1 {
+    pub fn get_domain(&self) -> String {
+        self.domain_name.clone()
+    }
+
+    pub fn get_subdomain(&self) -> String {
+        self.subdomain_name.get_string().unwrap_or_default()
+    }
```

Comment:
if you're just returning the objects without any modification, why the wrappers?

Reply:
Subdomain can be null. It's an OptionalString so we need to unwrap it.